integrations/googledrive/src/client.ts

```diff
@@ -0,0 +1,313 @@
+    const file = await this._fetchFile(fileId)
+    if (file.type !== 'normal') {
+      throw new RuntimeError(`Attempted to download a file of type ${file.type}`)
+    }
+
+    const { mimeType: contentType, size } = file
+    const fileDownloadResponse = await this._googleClient.files.get(
+      {
+        fileId,
+        alt: 'media',
+      },
+      {
+        responseType: 'stream',
+      }
+    )
+    const fileDownloadStream = fileDownloadResponse.data
+    const { file: bpFile } = await this._client.upsertFile({
+      key: fileId,
+      size,
+      contentType,
+      index,
+    })
+    const { id: bpFileId, uploadUrl } = bpFile
+
+    const headers = {
+      'Content-Type': contentType,
+      'Content-Length': size,
+    }
+    await axios
+      .put(uploadUrl, fileDownloadStream, {
+        maxBodyLength: size,
+        headers,
+      })
+      .catch((reason: AxiosError) => {
+        throw new RuntimeError(`Error uploading file to botpress file API: ${reason}`)
+      })
+
+    return {
+      bpFileId,
+    }
+  }
+
+  /**
+   * Removes internal fields and adds computed attributes
+   */
+  private async _getCompleteFileFromBaseFile(file: BaseNormalFile, optionalFilesCache?: FilesCache): Promise<File> {
+    const genericFile = convertNormalFileToGeneric(file)
+    const { filesCache } = await this._addParentsToFilesCache(genericFile, optionalFilesCache)
+    const { id, mimeType, name, parentId, size } = file
+    return {
+      id,
+      mimeType,
+      name,
+      parentId,
+      size,
+      path: this._getFilePath(id, filesCache),
+    }
+  }
+
+  /**
+   * Removes internal fields and adds computed attributes
+   */
+  private async _getCompleteFolderFromBaseFolder(
+    file: BaseFolderFile,
+    optionalFilesCache?: FilesCache
+  ): Promise<Folder> {
+    const genericFile = convertFolderFileToGeneric(file)
+    const { filesCache } = await this._addParentsToFilesCache(genericFile, optionalFilesCache)
+    const { id, mimeType, name, parentId } = file
+    return {
+      id,
+      mimeType,
+      name,
+      parentId,
+      path: this._getFilePath(id, filesCache),
+    }
+  }
+
+  private async _fetchFilesAndUpdateCache({
+    filesCache,
+    nextToken,
+    searchQuery,
+  }: {
+    filesCache: FilesCache
+    nextToken?: string
+    searchQuery?: string
+  }): Promise<{
+    newFilesIds: string[]
+    nextToken?: string
+  }> {
+    const listResponse = await this._googleClient.files.list({
+      corpora: 'user',
+      fields: GOOGLE_API_FILELIST_FIELDS,
+      q: searchQuery,
+      pageToken: nextToken,
+      pageSize: PAGE_SIZE,
+    })
+
+    const newNextToken = listResponse.data.nextPageToken ?? undefined
+    const unvalidatedDriveFiles = listResponse.data.files
+    if (!unvalidatedDriveFiles) {
+      throw new RuntimeError('No files were returned by the API')
+    }
+    const files = validateGenericFiles(unvalidatedDriveFiles)
```

Comment:
**nitpick:** this function does more than validation; I'd name it `parseGenericFiles`